File: practice/20251013/singleton.go
package main

import (
	"fmt"
	"sync"
)

type Singleton struct {
}

var once sync.Once
var instance *Singleton

func GetInstance() *Singleton {
	once.Do(func() {
		instance = &Singleton{}
		fmt.Println("Singleton created")
	})

	return instance
}

File: practice/20250929/Singleton.cpp


class Singleton {
private:
    Singleton() {}
public:
    Singleton(const Singleton&) = delete;
    Singleton& operator=(const Singleton&) = delete;

    static Singleton& getInstance() {
        static Singleton instance;
        return instance;
    }
}


class EagerSingleton {
private:
    EagerSingleton() {}

    static EagerSingleton instance;
public:
    EagerSingleton(const EagerSingleton&) = delete;
    EagerSingleton& operator=(const EagerSingleton&) = delete;

    static EagerSingleton& getInstance() {
        return instance;
    }
}

EagerSingleton EagerSingleton::instance;
